sklearn/decomposition/tests/test_fastica.py

```diff
@@ -223,34 +253,54 @@ def test_non_square_fastica(add_noise):
 
     # Check that we have estimated the original sources
     if not add_noise:
-        assert_almost_equal(np.dot(s1_, s1) / n_samples, 1, decimal=3)
-        assert_almost_equal(np.dot(s2_, s2) / n_samples, 1, decimal=3)
+        assert_allclose(np.dot(s1_, s1) / n_samples, 1, atol=1e-3)
+        assert_allclose(np.dot(s2_, s2) / n_samples, 1, atol=1e-3)
 
 
-def test_fit_transform():
+def test_fit_transform(global_random_seed, global_dtype):
     """Test unit variance of transformed data using FastICA algorithm.
 
     Check that `fit_transform` gives the same result as applying
     `fit` and then `transform`.
 
     Bug #13056
     """
-    rng = np.random.RandomState(0)
-    X = rng.random_sample((100, 10))
+    # multivariate uniform data in [0, 1]
+    rng = np.random.RandomState(global_random_seed)
+    X = rng.random_sample((100, 10)).astype(global_dtype)
+    max_iter = 300
     for whiten, n_components in [["unit-variance", 5], [False, None]]:
         n_components_ = n_components if n_components is not None else X.shape[1]
 
-        ica = FastICA(n_components=n_components, whiten=whiten, random_state=0)
-        Xt = ica.fit_transform(X)
+        ica = FastICA(
+            n_components=n_components, max_iter=max_iter, whiten=whiten, random_state=0
+        )
+        with warnings.catch_warnings():
+            # make sure that numerical errors do not cause sqrt of negative
+            # values
+            warnings.simplefilter("error", RuntimeWarning)
+            # XXX: for some seeds, the model do not converge. However this is not
+            # what we test here.
+            warnings.simplefilter("ignore", ConvergenceWarning)
+            Xt = ica.fit_transform(X)
         assert ica.components_.shape == (n_components_, 10)
-        assert Xt.shape == (100, n_components_)
+        assert Xt.shape == (X.shape[0], n_components_)
 
-        ica = FastICA(n_components=n_components, whiten=whiten, random_state=0)
-        ica.fit(X)
-        assert ica.components_.shape == (n_components_, 10)
-        Xt2 = ica.transform(X)
+        ica2 = FastICA(
+            n_components=n_components, max_iter=max_iter, whiten=whiten, random_state=0
+        )
+        with warnings.catch_warnings():
+            # make sure that numerical errors do not cause sqrt of negative
+            # values
+            warnings.simplefilter("error", RuntimeWarning)
+            warnings.simplefilter("ignore", ConvergenceWarning)
+            ica2.fit(X)
+        assert ica2.components_.shape == (n_components_, 10)
+        Xt2 = ica2.transform(X)
 
-        assert_array_almost_equal(Xt, Xt2)
+        # XXX: we have to set atol for this test to pass for all seeds when
+        # fitting with float32 data. Is this is a revealing a bug?
+        assert_allclose(Xt, Xt2, atol=1e-6 if global_dtype == np.float32 else 0.0)
```

Comment:
I noticed that if we use Student-t distributed data instead of uniformly distributed data in this test, then it's possible to make the test pass with `rtol=1e-3` (but not `rtol=1e-4`) and with a strict `atol=0.0`.

Also, I have check that there is not exact zeros in `Xt2` when assertion with the default tols fail (both with uniform and Student-t data).

Reply:
> To me, this case shows that atol really should depend on the way data is generated. Personally, I do not want to get done this rabbit hole.

`atol` does depend on the data ! it basically means "I consider elements less than atol to be zero". If you have data with an average magnitude of 1, you might want set atol to 1e-5 or something like that. If you have data with an average magnitude of 1e-12 (m.e.g data for instance), then you want to set atol to 1e-17 or so.